File: Java/src/main/java/Solution301_400/Solution350.java
package Solution301_400;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;

/**
 * @author windows
 */
public class Solution350 {
    public int[] intersect(int[] nums1, int[] nums2) {
        HashMap<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < nums1.length; i++) {
            map.put(nums1[i], map.getOrDefault(nums1[i], 0) + 1);
        }
        ArrayList<Integer> list = new ArrayList<>();
        for (int i = 0; i < nums2.length; i++) {
            if(map.containsKey(nums2[i]) && map.get(nums2[i]) > 0) {
                list.add(nums2[i]);
                map.put(nums2[i], map.get(nums2[i]) - 1);
            }
        }
        return Arrays.stream(list.toArray(new Integer[list.size()])).mapToInt(Integer::intValue).toArray();
    }


    public static void main(String[] args) {
        Solution350 solution350 = new Solution350();
    }
}

File: C++/src/Solution301_400/Solution350.cpp
//
// Created by windows on 2023/3/2.
//

#include "Solution350.h"
#include <functional>
#include <iostream>
#include <vector>
#include <unordered_set>
#include <unordered_map>
#include <string>
using namespace std;
#include <unordered_map>


vector<int> Solution350::intersect(vector<int> &nums1, vector<int> &nums2) {
    unordered_map<int, int> hashtable;
    vector<int> res;
    for (int i = 0; i < nums1.size(); i++) {
        hashtable[nums1[i]] = hashtable[nums1[i]] + 1;
    }
    for (int i = 0; i < nums2.size(); ++i) {
        if (hashtable.count(nums2[i]) && hashtable[nums2[i]] > 0) {
            res.push_back(nums2[i]);
            hashtable[nums2[i]] = hashtable[nums2[i]] - 1;
        }
    }
    return res;
}
